File: typescript/src/calisthenics-exercises/small-entities.ts
export class ReportService {
  private cache: Map<string, string> = new Map()

  generateCsvReportFromJson(jsonInput: string, delimiter: string = ','): string {
    let data: unknown
    try {
      data = JSON.parse(jsonInput)
    } catch (e) {
      throw new Error('Invalid JSON')
    }

    if (!Array.isArray(data)) {
      throw new Error('Expected array')
    }

    const headers = Object.keys(data[0] as any)
    const lines = [headers.join(delimiter)]
    for (const row of data as Array<Record<string, any>>) {
      const values = headers.map((h) => String(row[h] ?? ''))
      lines.push(values.join(delimiter))
    }
    const result = lines.join('\n')

    this.cache.set('last', result)
    return result
  }
}

File: csharp/src/calisthenics-exercises/SmallEntities.cs
namespace RefactoringAvanzado.CalisthenicExercises;

public class ReportService
{
    private Dictionary<string, string> cache = new Dictionary<string, string>();

    public string GenerateCsvReportFromJson(string jsonInput, string delimiter = ",")
    {
        dynamic data;
        try
        {
            data = System.Text.Json.JsonSerializer.Deserialize<object>(jsonInput);
        }
        catch
        {
            throw new Exception("Invalid JSON");
        }

        if (data is not System.Text.Json.JsonElement element || element.ValueKind != System.Text.Json.JsonValueKind.Array)
        {
            throw new Exception("Expected array");
        }

        var array = element.EnumerateArray().ToList();
        if (array.Count == 0)
        {
            return string.Empty;
        }

        var firstRow = array[0];
        var headers = firstRow.EnumerateObject().Select(p => p.Name).ToList();
        var lines = new List<string> { string.Join(delimiter, headers) };

        foreach (var row in array)
        {
            var values = headers.Select(h =>
            {
                if (row.TryGetProperty(h, out var prop))
                {
                    return prop.ToString();
                }
                return string.Empty;
            });
            lines.Add(string.Join(delimiter, values));
        }

        var result = string.Join("\n", lines);
        cache["last"] = result;
        return result;
    }
}
